modules/tensor_mechanics/src/actions/TensorMechanicsAction.C

```diff
@@ -451,6 +470,44 @@ TensorMechanicsAction::actOutputGeneration()
   }
 }
 
+void
+TensorMechanicsAction::actEigenstrainNames()
+{
+  // Determine all the materials(eigenstrains) all ready created
+  auto materials = _problem->getMaterialWarehouse().getObjects();
+
+  for (auto & mat : materials)
+  {
+    const InputParameters & params = mat->parameters();
+    // Check for eigenstrain names, only deal with those materials
+    if (params.isParamValid("eigenstrain_name"))
+    {
+      // It might be possible that an eigenstrain is applied multiple blocks
+      auto block = params.get<std::vector<SubdomainName>>("block");
+      auto name = params.get<std::string>("eigenstrain_name");
+      for (unsigned int i = 0; i < _subdomain_names.size(); ++i)
+      {
+        for (unsigned int j = 0; j < block.size(); ++j)
+        {
+          auto unique_name = std::find(
+              _block_based_eigenstrain_names.begin(), _block_based_eigenstrain_names.end(), name);
+          // For only 1 block, the names shouldn't be duplicate
+          if (block.at(j) == _subdomain_names.at(i) && _subdomain_names.size() < 2)
+            _block_based_eigenstrain_names.push_back(name);
```

Comment:
Where are you recording which block the eigenstrain name applies to? You shoud use 

```
std::map<SubdomainID, std::set<std::string>>
```

Also, look up the subdomain IDs rather than the names. And use `BlockRestrictable::blockIDs()` rather than querying the `blocks` parameter.